### PR TITLE
ARROW-7169: [C++] Vendor uriparser library

### PR DESCRIPTION
**Pull request by pitrou**
The library is only used internally (the types and headers are not exposed).
Vendoring it makes it easier for core functionality to depend on it.

### CODE REVIEW

**Comment by pitrou** [comment]
@kou Do you have any concerns?

File: cpp/src/arrow/vendored/uriparser/UriConfig.h.in [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
```
Comment (kou) [inline cpp/src/arrow/vendored/uriparser/UriConfig.h.in]: We don't need this file. Because we bundle generated `config.h`.

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/vendored/uriparser/UriConfig.h.in (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/vendored/uriparser/UriConfig.h.in]: True. I'll remove it.

**Review by kszucs — APPROVED** [review]
LGTM

**Comment by kszucs** [comment]
@pitrou thanks for updating the packaging builds as well

**Comment by pitrou** [comment]
@kszucs Yes, I had forgotten to update some of them for the double-conversion vendoring.

**Comment by kszucs** [comment]
Our nightlies are in pretty bad shape, so it would not make much sense to run them. Merging.

**Pull request closed without merging**